server/src/resources/base.validator.js

```diff
@@ -1,12 +1,81 @@
-module.exports = async function validateRequest(ctx, validateFunction) {
+const Joi = require('joi');
+
+const joiOptions = {
+  abortEarly: false,
+};
+
+/**
+ * Parse and return list of errors
+ * @param {object} joiError
+ * @return {object[]}
+ */
+const parseJoiErrors = (joiError) => {
+  let resultErrors = [];
+  if (joiError && joiError.details instanceof Array) {
+    resultErrors = joiError.details.map((error) => {
+      const pathLastPart = error.path.slice(error.path.length - error.context.key.length);
+
+      if (pathLastPart === error.context.key) {
+        return { [error.path]: error.message };
+      }
+
+      return { [error.context.key]: error.message };
+    });
+  }
+
+  return resultErrors;
+};
+
+/**
+ *  Validate request and send 400(bad request), when request is not valid
+ * @param {object} ctx
+ * @param {object|Function} schema - Joi validation schema or validation function
+ * @param {Function} validateFn
+ * @return {object}
+ */
+module.exports = async (ctx, schema, validateFn) => {
+  let validateFunc = validateFn;
+  let joiSchema = schema;
+
+  if (typeof schema === 'function') {
+    validateFunc = schema;
+    joiSchema = null;
+  }
+
+  let joiResult;
+  if (joiSchema) {
+    let body;
+    if (ctx.method === 'GET') {
+      body = ctx.params;
+    } else {
+      body = ctx.request.body;
+    }
+
+    joiResult = Joi.validate(body, joiSchema, joiOptions);
+    if (joiResult.error) {
+      ctx.errors = parseJoiErrors(joiResult.error);
+
+      ctx.status = 400;
+      ctx.body = { errors: ctx.errors };
+      return { errors: ctx.errors };
+    }
+  }
+
   ctx.errors = [];
-  const data = await validateFunction(ctx) || {};
-  data.isValid = !ctx.errors.length;
+  const data = await validateFunc(joiResult.value);
```

Comment:
`joiResult` can be `undefined` if call validator without schema.
And `ctx.status` and `ctx.body` should be set in controller.